### PR TITLE
Added search-depth-limit parameter

### LINKED ISSUE #18: Error: Maximum call stack size exceeded
When a project has too many files `getAllCycles` eventually will hit the call stack size limit:

```
The 'no-circular-imports' rule threw an error in 'file.tsx':
RangeError: Maximum call stack size exceeded
    at NoCircularImportsWalker.getAllCycles (/Users/user/project/node_modules/tslint-no-circular-imports/noCircularImportsRule.js:195:30)
    at NoCircularImportsWalker.getAllCycles (/Users/user/project/node_modules/tslint-no-circular-imports/noCircularImportsRule.js:193:30)
    at NoCircularImportsWalker.getAllCycles (/Users/user/project/node_modules/tslint-no-circular-imports/noCircularImportsRule.js:193:30)
    at NoCircularImportsWalker.getAllCycles (/Users/user/project/node_modules/tslint-no-circular-imports/noCircularImportsRule.js:193:30)
    at NoCircularImportsWalker.getAllCycles (/Users/user/project/node_modules/tslint-no-circular-imports/noCircularImportsRule.js:193:30)
    at NoCircularImportsWalker.getAllCycles (/Users/user/project/node_modules/tslint-no-circular-imports/noCircularImportsRule.js:193:30)
    at NoCircularImportsWalker.getAllCycles (/Users/user/project/node_modules/tslint-no-circular-imports/noCircularImportsRule.js:193:30)
    at NoCircularImportsWalker.getAllCycles (/Users/user/project/node_modules/tslint-no-circular-imports/noCircularImportsRule.js:193:30)
    at NoCircularImportsWalker.visitSourceFile (/Users/user/project/node_modules/tslint-no-circular-imports/noCircularImportsRule.js:102:34)
    at NoCircularImportsWalker.SyntaxWalker.visitNode (/Users/user/project/node_modules/tslint/lib/language/walker/syntaxWalker.js:482:22)
```

I've tried to refactor it so it will be asynchronus but it ran out of memory 😞 
Is there any way to make it work with large projects?

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Limits the depth of cycle reporting to a fixed to a fixied size limit for a list of files.
This helps improve performance, as most cycles do not surpass a few related files.

As I mentioned in the issue, we could theoretically fix the crashes by switching to an iterative solution. This is still a desirable change IMO because erroring files will do so faster.

Fixes #18 (recreation of #20 but with a good git history)

### CODE REVIEW

File: noCircularImportsRule.ts [hunk 1]
```diff
@@ -113,6 +136,26 @@ function walk(context: Lint.WalkContext<Options>) {
...
   }
+
+  function getAllCycles(moduleName: string, accumulator: string[] = [], iterationDepth = 0): string[][] {
+    const moduleImport = imports.get(moduleName)
+    if (!moduleImport) return []
+    if (accumulator.indexOf(moduleName) !== -1)
+      return [accumulator]
+
+    if (iterationDepth >= context.options.searchDepthLimit)
+      return []
```
Comment (bcherny) [inline noCircularImportsRule.ts]: Worth a `console.warn`?

Comment (JoshuaKGoldberg) [reply noCircularImportsRule.ts]: Ideally no. Large codebases will hit this a lot.

Comment (bcherny) [reply noCircularImportsRule.ts]: With this change, when I have a cycle in a big codebase, will I ever find out about it?

Comment (bcherny) [reply noCircularImportsRule.ts]: Or do you think the log wouldn't be useful, since there likely isn't a cycle (just deep dep graph?

Comment (JoshuaKGoldberg) [reply noCircularImportsRule.ts]: Yeah that's what I'm thinking. From working in code bases with a few hundred or a few thousand files, most of them just import a few around their own directory. Even `50` is pretty extreme.

...then again... you're right that if this results in nothing, but we know there was at least one cycle and it happens to be ridiculously long, I don't know what we could do intelligently show that.

Sorry for the brevity, on mobile :)

File: test/case-long-stack/two.ts [hunk 2]
```diff
@@ -0,0 +1 @@
+import * as two from "./two";
```
Comment (bcherny) [inline test/case-long-stack/two.ts]: What are you trying to test with these files?

Comment (JoshuaKGoldberg) [reply test/case-long-stack/two.ts]: These validate that the `search-depth-limit` option is respected.

`one` imports from `two` imports from `three` imports from `four` imports from `one` for a depth of four. Adding the option to `tslint.json` means these shouldn't cause a lint complaint, which is why I didn't modify the list of expected complaints.

**Comment by bcherny** [comment]
Looks good, sorry about the delay. Published 0.6.0.

**Comment by JoshuaKGoldberg** [comment]
Awesome, thanks!

**Pull request merged**